File: 239 Sliding Window Maximum/solution.py
class Solution:
    # @param {integer[]} nums
    # @param {integer} k
    # @return {integer[]}
    def maxSlidingWindow(self, nums, k):
        helper = collections.deque()
        result = []
        for i in range(len(nums)):
            while helper and nums[i]>=nums[helper[-1]]:
                helper.pop()
            helper.append(i)
            if helper[0]==i-k:
                helper.popleft()
            if i>=k-1:
                result.append(nums[helper[0]])
        return result
        

File: 239 Sliding Window Maximum/solution.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        if (k <= 0) return {};
        if (k == 1) return nums;
        vector<int> result;
        deque<int> helper;
        for (int i = 0; i < nums.size(); i++) {
            if (helper.empty()) {
                helper.push_back(nums[i]);
                continue;
            }
            if ((i >= k) && (nums[i - k] == helper.front())) {
                helper.pop_front();
            }
            while (!helper.empty() && helper.back() < nums[i]) {
                helper.pop_back();
            }
            helper.push_back(nums[i]);
            if (i >= k - 1) {
                result.push_back(helper.front());
            }
        }
        return result;
    }
};



